### PR TITLE
Convert to a more modular, extensible process for generating data.

### PR DESCRIPTION
**Pull request by stuartforrest-infinity**
This PR fundamentally changes the structure, configuration and invocation of the step function as it exists in master.

In looking to use this to write data to an RDS I wanted a degree of flexibility and independence between handling different storage targets within a single step function and have built out this as a concept. The intention is for it to be extensible with relative ease to support further targets or processing data in different ways for existing targets.

I have added a `Readme` to explain how it works and is configured. Have a read there for what the step function now looks like.

### CODE REVIEW

File: serverless.yml [hunk 1]
```diff
@@ -7,85 +7,111 @@ plugins:
...
+      #       # This runs at 8PM every day - amend as required
+      #       rate: cron(0 20 * * ? *)
+      #       input:
+      #         # Bucket is required - there is no default
+      #         # Cannot use CloudFormation function here - including Sub
+      #         bucket: ${self:service}-${opt:stage, self:provider.stage}-cafe-configs
+      #         # Defaults to storeConfig.yaml
+      #         key: storeConfig.yaml
+      #         # Defaults to 0 - all stores
+      #         store_count: 1
```
Comment (lewisrichardson) [inline serverless.yml:27]: Will this be needed at any point or is it commented in case it needs to be used for something?

File: serverless.yml [hunk 2]
```diff
@@ -101,53 +127,54 @@ provider:
...
-    - Effect: Allow
-      Action:
-        - kinesis:PutRecords
-      Resource:
-        - !GetAtt StoreStream.Arn
+    # - Effect: Allow
+    #   Action:
+    #     - kinesis:PutRecords
+    #   Resource:
+    #     - !GetAtt StoreStream.Arn
```
Comment (lewisrichardson) [inline serverless.yml:134]: Is this needed

File: serverless.yml [hunk 3]
```diff
@@ -101,53 +127,54 @@ provider:
...
-      Type: AWS::S3::Bucket
-      Properties:
-        BucketName: ${self:service}-${opt:stage, self:provider.stage}-cafe-configs-2
\ No newline at end of file
+    # StoreStream:
+    #   Type: AWS::Kinesis::Stream
+    #   Properties:
+    #       Name: store-stream
+    #       RetentionPeriodHours: 168
+    #       ShardCount: 1
```
Comment (lewisrichardson) [inline serverless.yml:180]: Is this needed

File: src/lambdas/generate_stream_data/handler.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        records = []
+        for record in transaction_data:
+            records.append(
+                {
+                    'Data': bytes(json.dumps(record), 'utf-8'),
+                    'PartitionKey': store.name,
+                }
+            )
+            
+        chunked_data = chunks(records, 500)
```
Comment (lewisrichardson) [inline src/lambdas/generate_stream_data/handler.py:63]: Magic number? 😄

File: src/lambdas/generate_stream_data/handler.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+def make_handler():
+    def handle(event, context):
+        store = event.get('store')
+        streams = event.get('streams')
+
+        store = build_store(store)
+        transactions = generate_transactions(store)
+        transaction_data = create_transaction_data(transactions)
+        #today = datetime.datetime.today()
```
Comment (lewisrichardson) [inline src/lambdas/generate_stream_data/handler.py:52]: Needed?

File: src/lambdas/load_config/handler.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            Bucket=config_bucket, Key=key)
+        data = yaml.load_all(resp['Body'], Loader=yaml.FullLoader)
+        stores = list(data)
+
+        explicit_count_configs = [
+            config for config in targets if config.get('store_count', 0) != 0]
+        spread_count_configs = [
+            config for config in targets if config.get('store_count', 0) == 0]
+
+        ret = {}
```
Comment (lewisrichardson) [inline src/lambdas/load_config/handler.py:31]: Is this a descriptive variable name?

File: src/lambdas/validate_config/handler.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        config_bucket = event.get('bucket')
+        key = event.get('key')
+
+        if not config_bucket:
+            raise Exception("config_bucket must be set - there is no default")
+        if not key:
+            key = event["key"] = DEFAULT_STORE_CONFIG_FILEPATH
+        
+        print(event)
```
Comment (lewisrichardson) [inline src/lambdas/validate_config/handler.py:60]: Needed?

**Review by lewisrichardson — APPROVED** [review]

File: src/lambdas/load_config/handler.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            # TODO: Check if there are any stores?
+            # TODO: Floor here will leave some stores spare - track index in loop and on last allocate all remaining stores
+            store_count = math.floor(
+                (remaining_stores/len(spread_count_configs)) - 1)
+            for target_config in spread_count_configs:
+                ret[target_config["type"]] = {
+                    **target_config,
+                    "stores": stores[0:store_count]
+                }
+                del stores[0:store_count]
```
[the remark this answers is not in the document]
Comment (stuartforrest-infinity) [reply src/lambdas/load_config/handler.py:52]: I tend be more WET than DRY!

**Pull request merged**